V0.5.0-Implement String Localization and Hardcoded String Detection

cicerohellmann opened the pull request:
## 🚀 Automated String Localization Enforcement with Custom Ktlint Rules

### 📋 Overview
This PR implements a **custom Ktlint rule system** to automatically detect and flag hardcoded strings in Composable functions, replacing manual testing with automated code quality enforcement during the build process.

### ✨ What's New

#### 🔧 **New Module: `ktlint-rules/`**
- **Custom Rule Engine**: Implements `NoHardcodedStringsRule` that intelligently detects user-visible hardcoded strings
- **Smart Filtering**: Automatically excludes technical strings (URLs, log tags, SQL, date formats, etc.)
- **Preview-Aware**: Skips `@Preview` functions since they contain legitimate test data
- **Service Provider Integration**: Uses SPI pattern for seamless Ktlint discovery

#### 🎯 **Key Features**
- **Automated Detection**: Runs during build process, no manual intervention needed
- **Context-Aware**: Only flags strings in `@Composable` functions that are likely user-visible
- **Developer-Friendly**: Clear error messages with localization guidance
- **Build Integration**: Seamlessly integrates with existing Ktlint setup

### 📁 Files Changed

#### ➕ **Added**
- `ktlint-rules/build.gradle.kts` - Module configuration with Ktlint dependencies
- `ktlint-rules/src/main/kotlin/.../NoHardcodedStringsRule.kt` - Core rule implementation (131 lines)
- `ktlint-rules/src/main/kotlin/.../ComposeActorsRuleSetProvider.kt` - Rule provider registration
- `ktlint-rules/src/main/resources/META-INF/services/...` - SPI configuration for rule discovery

#### 🔄 **Modified**
- `app/build.gradle.kts` - Added `ktlintRuleset(project(":ktlint-rules"))` dependency
- `settings.gradle.kts` - Included new module with `include(":ktlint-rules")`
- `.gitignore` - Enhanced to properly ignore IDE-specific files

#### ❌ **Removed**
- `app/src/test/.../HardcodedStringTest.kt` - Replaced manual JUnit test with automated rule
- IDE configuration files (`.idea/compiler.xml`, `.idea/gradle.xml`, `.idea/misc.xml`) - Cleaned up repository

### 🧠 **Smart String Detection Logic**

The rule intelligently filters strings using multiple criteria:

```kotlin
// ✅ Will be flagged (user-visible)
Text("Welcome to our app")
Button(onClick = {}) { Text("Click me") }

// ❌ Won't be flagged (technical/excluded)
Log.d("TAG", "Debug message")
Image(contentDescription = "cd_profile_image")
Text("https://api.example.com/users")
```

### 🔄 **Migration Benefits**

| Before | After |
|--------|-------|
| Manual JUnit test execution | Automatic build-time checking |
| Requires developer to remember to run tests | Integrated into CI/CD pipeline |
| Static test cases | Dynamic analysis of all Composable functions |
| Limited to predefined test scenarios | Comprehensive codebase scanning |

### 🚦 **Usage**

The rule automatically runs with your existing Ktlint setup:

```bash
# Run Ktlint check (includes custom rule)
./gradlew ktlintCheck

# Auto-format code (custom rule provides guidance only)
./gradlew ktlintFormat
```

### 🎉 **Impact**

- **Improved Code Quality**: Automatic enforcement of localization best practices
- **Developer Experience**: Clear, actionable feedback during development
- **Consistency**: Uniform string localization standards across the project
- **CI/CD Ready**: Integrates seamlessly with existing build pipelines

This implementation transforms string localization from a manual checklist item into an automated quality gate, ensuring consistent internationalization practices across the entire Compose Actors codebase! 🌍✨

cicerohellmann committed:
Add test to detect hardcoded strings in Composables

Introduces HardcodedStringTest to scan Kotlin source files for hardcoded user-visible strings within @Composable functions, excluding preview and technical strings. This helps enforce best practices for string localization and maintainability.

cicerohellmann committed:
Localize hardcoded strings in UI components

Replaced hardcoded UI strings with string resources across multiple screens to improve localization support. Added new string resources for tab titles, category labels, error messages, and content descriptions.

cicerohellmann committed:
Update HardcodedStringTest.kt

cicerohellmann committed:
Add custom ktlint rule for hardcoded strings in Compose

Introduces a new :ktlint-rules module with a custom rule to detect hardcoded user-visible strings in @Composable functions, replacing the previous test-based approach. Updates .gitignore for new build outputs and removes obsolete IntelliJ project files. Integrates the new rule set into the app module and updates settings.gradle.kts to include the new module.

RajashekarRaju reviewed (commented):
The rule works pretty well for most of the cases.
I have found below things which are worth looking at.

### 1. Companion‐object constant

```kotlin
data class HomeOptionItems(
    val id: Int,
    val title: Int,
    val icon: Int,
) {
    companion object {
        // Did not catch
        private const val HURRY = "LULU"
    }
}
```

This is a private constant in a companion object and not directly user-facing UI text, so it’s reasonable to leave it out of our hard-coded-string rule.

---

### 2. Private constant in a UI file

```kotlin
// Did not catch
private const val HAH = "LULU"
```

Currently our rule only scans inside `@Composable` functions, so top-level or file-level constants are skipped. Should we extend it to flag raw `const val` declarations in UI files as well? If these constants ever end up in the UI, we’d miss them.
**What do you think ?**

---

### 3. Local `val` initializer in a Composable

```kotlin
@Composable
fun AboutScreen(...) {
    // Did not catch
    val userName = "Lulu"
    …
}
```

This variable holds a hard-coded name that’s likely shown on screen. We should catch any string literal initializers inside Composables I believe.

---

### 4. Mixed template literal

```kotlin
CaTextBody1(
    text = "$userName logged in",
    modifier = Modifier,
)
```

Even though it contains interpolation, the literal fragment `" logged in"` is hard-coded. Our rule currently skips templates with `${...}`.

---

### 5. Custom button label in AlertDialog

```kotlin
AlertDialog(
    …,
    confirmButton = {
        CaTextButton(
            text = "Dismiss",
            modifier = Modifier,
        )
    }
)
```
---

### 6. Empty contentDescription

```kotlin
LoadNetworkImage(
    …,
    contentDescription = "",
    …
)
```

An explicit empty string, should we allow these?

cicerohellmann committed:
Clean up project files and improve NoHardcodedStringsRule

Removed .idea configuration files and updated .gitignore to exclude .idea and .kotlin directories. Enhanced the NoHardcodedStringsRule to better detect hardcoded strings in both functions and properties, including improved handling of template strings and property declarations. Also removed unnecessary test dependencies from ktlint-rules/build.gradle.kts and fixed a typo in the README.

cicerohellmann committed:
Refactor to avoid hardcoded email in sign-up preview

Replaces the hardcoded email in SignUpScreenAwaitingConfirmationStepPreview with a string resource. Adds 'sample_email' to strings.xml and improves template string detection in NoHardcodedStringsRule for better linting of hardcoded strings.

cicerohellmann committed:
Remove unused sample_email string and hardcode preview email

Replaces the use of the sample_email string resource with a hardcoded email in the SignUpScreen preview and removes the unused sample_email entry from strings.xml.

cicerohellmann committed:
Replace hardcoded 'Dismiss' with string resource

Updated the ShowAlertDialog component to use a string resource for the 'Dismiss' button instead of a hardcoded value. Added the 'dismiss' string to strings.xml. Improved NoHardcodedStringsRule to better detect technical strings and domain/date patterns. Registered custom Ktlint ruleset provider.

cicerohellmann committed:
Replace hardcoded strings with string resources

Refactored UI components to use stringResource for user-facing text, improving localization support. Added new string resources for error messages and feature announcements. Updated the NoHardcodedStringsRule to refine technical string detection.

cicerohellmann committed:
Merge branch 'development' into ktlint-localization

cicerohellmann committed:
Update UiStateHandler.kt
